File: src/chipschallenge/blockreactions/FlippersTo.java
package chipschallenge.blockreactions;

import chipschallenge.Block;
import chipschallenge.Block.Type;
import chipschallenge.BlockContainerFullException;
import chipschallenge.Inventory.Boots;

/**
 * Move to flippers
 */
public class FlippersTo extends BlockReaction {

    private FlippersTo() {
    }
    private static FlippersTo mInstance = null;

    public static synchronized FlippersTo getInstance() {
        if (mInstance == null) {
            mInstance = new FlippersTo();
        }
        return mInstance;
    }

    // Chip can pick up flippers
    public void react(Block moving, Block standing) throws BlockContainerFullException {
        if (moving.isChip()) {
            takeBoots(Boots.FLIPPERS);
            standing.replace(createBlock(Type.FLOOR));
            //TODO: Play take-item sound
        }
    }

    // Only chip and blocks
    public boolean canMove(Block moving, Block standing) {
        return moving.isChip() || moving.isBlock();
    }
}

File: ChipsChallenge.Shared/Blockreactions/FlippersTo.cs
namespace ChipsChallenge.Shared.Blockreactions
{
    using Boots = Inventory.Boots;

    public class FlippersTo : NoSlipReaction
    {
        private static FlippersTo instance;

        private FlippersTo()
        {
        }

        public static FlippersTo Instance
        {
            get
            {
                lock (typeof(FlippersTo))
                {
                    if (instance == null)
                    {
                        instance = new FlippersTo();
                    }
                    return instance;
                }
            }
        }

        public override void React(Block moving, Block standing)
        {
            if (moving.Chip)
            {
                TakeBoots(Boots.FLIPPERS);
                standing.Replace(CreateBlock(Block.Type.FLOOR));
            }
        }

        public override bool canMove(Block moving, Block standing)
        {
            return moving.Chip || moving.IsBlock();
        }
    }
}